pennylane/shots.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the ShotAPI class to hold shot-related information."""
+
+from collections import namedtuple
+from collections.abc import Sequence
+from functools import singledispatchmethod
+import numpy as np
+
+ShotTuple = namedtuple("ShotTuple", ["shots", "copies"])
+
+
+class ShotAPI:
+    """A data class that stores shot information needed in order to execute tapes."""
+
+    # pylint:disable=too-few-public-methods
+
+    total_shots, shot_vector, shot_list = 0, [], []
+
+    @singledispatchmethod
+    def __init__(self, shots):
+        if shots is not None:
+            raise ValueError(
+                "Shots must be a single non-negative integer or a sequence of non-negative integers."
+            )
+
+    @__init__.register
+    def __int_init__(self, shots: int):
+        if shots < 1:
+            raise ValueError(f"The specified number of shots needs to be at least 1. Got {shots}.")
+        self.total_shots = shots
+        self.shot_list = [shots]
+        self.shot_vector = [ShotTuple(shots, 1)]
+
+    @__init__.register
+    def __list_init__(self, shots: Sequence):
```

Comment:
This is something that should've been done for the current implementation of `shots` in `Device`, but we should probably add an element-wise check to ensure that only positive integers are present in the shot list. It would be better for consistency as well as the check is already performed when `shots` is an integer.

Reply:
done. i got rid of the singledispatch since it wasn't really necessary and might make things harder in the future